src/Gui/TaskCSysDragger.cpp

```diff
@@ -39,139 +47,586 @@
+    auto rootPlacement = vp->getObjectPlacement();
+
+    auto selectedObjectPlacement = rootPlacement.inverse() * globalPlacement
+        * subObjectPlacementProvider->calculate(msg.Object, localPlacement);
+
+    auto label = msg.pOriginalMsg
+        ? QStringLiteral("%1#%2.%3")
+              .arg(QLatin1String(msg.pOriginalMsg->pObjectName),
+                   QLatin1String(msg.pObjectName),
+                   QLatin1String(msg.pSubName))
+        : QStringLiteral("%1.%2").arg(QLatin1String(msg.pObjectName),
+                                      QLatin1String(msg.pSubName));
+
+    switch (selectionMode) {
+        case SelectionMode::SelectTransformOrigin: {
+            if (msg.Type == SelectionChanges::AddSelection) {
+                ui->referenceLineEdit->setText(label);
+                customTransformOrigin = selectedObjectPlacement;
+                updateTransformOrigin();
+                setSelectionMode(SelectionMode::None);
+            } else {
+                vp->setTransformOrigin(selectedObjectPlacement);
+            }
+
+            break;
+        }
+
+        case SelectionMode::SelectAlignTarget: {
+            vp->setDraggerPlacement(rootPlacement * selectedObjectPlacement);
+
+            if (msg.Type == SelectionChanges::AddSelection) {
+                vp->updateTransformFromDragger();
+                vp->updatePlacementFromDragger();
+
+                setSelectionMode(SelectionMode::None);
+            }
+
+            break;
+        }
+
+        default:
+            // no-op
+            break;
+    }
+}
+
+void TaskTransform::onAlignRotationChanged()
+{
+    updateDraggerLabels();
+    updateTransformOrigin();
+}
+
+void TaskTransform::onAlignToOtherObject()
+{
+    setSelectionMode(SelectionMode::SelectAlignTarget);
+}
+
+void TaskTransform::onFlip()
+{
+    auto placement = vp->getDraggerPlacement();
+
+    placement.setRotation(placement.getRotation()
+                          * Base::Rotation::fromNormalVector(Base::Vector3d(0, 0, -1)));
+
+    vp->setDraggerPlacement(placement);
+
+    vp->updateTransformFromDragger();
+    vp->updatePlacementFromDragger();
+}
+
+void TaskTransform::onPickTransformOrigin()
+{
+    setSelectionMode(selectionMode == SelectionMode::None ? SelectionMode::SelectTransformOrigin
+                                                          : SelectionMode::None);
+}
+
+void TaskTransform::onPlacementModeChange(int index)
+{
+    placementMode = ui->placementComboBox->currentData().value<PlacementMode>();
+
+    updateTransformOrigin();
+}
+
+void TaskTransform::updateTransformOrigin()
+{
+    auto getTransformOrigin = [this](const PlacementMode& mode) -> Base::Placement {
+        switch (mode) {
+            case PlacementMode::ObjectOrigin:
+                return {};
+            case PlacementMode::Centroid:
+                if (const auto com = centerOfMassProvider->ofDocumentObject(vp->getObject())) {
+                    return Base::Placement { *com, {} };
```

Comment:
I don’t think you need to write `Base::Placement` here as you already defined the return type of the lambda
```suggestion
                    return { *com, {} };
```